### ISSUE TITLE
Create node / particle sets

### ISSUE THREAD

**Issue by kks32** [cb-geo/mpm#356]
In the mesh class, we should be able to create a set of nodes and particles on which loading or restraints can be applied. A code snipped on how to create a node set:

```
//! Assign node sets
//! \brief Assign loads and restrain node sets
//! \return create_status Return success or failure of node set creation
template <unsigned Tdim>
bool mpm::Mesh<Tdim>::assign_node_sets() {
  auto node_set = json_mesh_["node_sets"];
  bool set_status = true;
  if (node_set.size() == 0) set_status = false;

  // Iterate through JSON node_sets object
  for (auto it = node_set.begin(); it != node_set.end(); ++it) {
    auto nodes =
        this->read_node_id(this->json_mesh_["cwd"].template get<std::string>() +
                           it.value().template get<std::string>());

    // Remove duplicates
    std::sort(nodes.begin(), nodes.end());
    nodes.erase(std::unique(nodes.begin(), nodes.end()), nodes.end());

    if (!this->create_nodes_set(it.key(), nodes)) set_status = false;
  }
  return set_status;
}
```

to create a node_set:
```
//! Create node sets
//! \brief Create node sets to assign forces / restraings
//! \param[in] key Key to store vector of nodes in a map
//! \param[in] nodes List of node indices
//! \return create_status Return success or failure of node set creation
template <unsigned Tdim>
bool lem::Mesh<Tdim>::create_nodes_set(const std::string& key,
                                       const std::vector<unsigned>& nodes) {

  bool node_set_create = false;
  std::vector<std::shared_ptr<lem::SolidNode<Tdim>>> set_solid_nodes;

  try {
    // Parse node index files
    for (const auto& nodeid : nodes) {
      // Search if node id is in vector of solid nodes
      auto node_itr = solid_nodes_.at(nodeid);
      // If a solid node pointer is found, push to a new set of nodes
      set_solid_nodes.push_back(node_itr);
    }
  } catch (std::exception& except) {
    console_->error("Create node set: {}", except.what());
  }

  if (set_solid_nodes.size() > 0) {
    node_set_.emplace(key, set_solid_nodes);
    node_set_create = true;
    console_->info("Node set created with: {}", key);
  } else
    node_set_create = false;

  return node_set_create;
}

```

In addition we need to create an iterate function to iterate over these node and particle sets:
```
//! Iterate over node set using a given key
//! \param[in] key Key of node_set
//! \tparam Toper operator on each node
template <unsigned Tdim>
template <typename Toper>
void mpm::Mesh<Tdim>::iterate_over_node_set(const std::string& key,
                                            Toper oper) {
  tbb::parallel_for_each(node_set_.at(key).begin(), node_set_.at(key).end(),
                         oper);
}
```

Node set is implemented as:
```
  //! Map of nodes to string to create sets
  std::map<std::string, std::vector<std::shared_ptr<lem::SolidNode<Tdim>>>>
      node_set_;
```

**Comment by kks32** [comment]
Additional features can be found in LEM for node_set: 

https://github.com/cb-geo/lem/blob/develop/include/mesh.h

https://github.com/cb-geo/lem/blob/develop/include/mesh.tcc

**Comment by kks32** [comment]
Fixed in #401
